### ISSUE TITLE
Add Uploads API

### ISSUE THREAD

**Issue by sgillies** [mapbox/mapbox-sdk-py#14]
I feel like the most useful thing this module can do for a user is abstract over the staging requirement https://www.mapbox.com/developers/api/uploads/#Stage.a.file.on.Amazon.S3.

Would be nice if this fit well with Rasterio (raster datasets as a native concept). Shouldn't depend on Rasterio.

**Comment by perrygeo** [comment]
The upload API currently requires 4 steps to completion

1. get credentials
2. s3 cp the data
3. initiate import 
4. poll for completion status

How much abstraction do we want here? The Javascript sdk has the user write their own s3 upload code as a callback. https://github.com/mapbox/mapbox-sdk-js/blob/master/lib/services/uploads.js#L77

This would certainly be an option for the python sdk too. But your description hints at doing a few of the steps (1, 2 and possibly 3) via a single, synchronous, high-level API call. @sgillies what are your thoughts on this?

.. Then of course we have list and delete but those seem a bit easier as they are quick, synchronous, single steps.

**Comment by sgillies** [comment]
@perrygeo before this project began, I'd been thinking about writing an upload plugin for Rasterio's CLI. Something like a `rio mbx-upload example.tif`. I think that still makes sense for some kinds of raster data where Rasterio features can improve upload performance (upload of VRTs, for example, or multipart uploads of image tiles). If we didn't abstract over steps 1-3 in this project (sdk-py) we'd have to do it in the plugin. And if we want to expose uploads in this project's CLI, the same abstraction will be useful.

**Comment by perrygeo** [comment]
Yes, I'm leaning towards exposing all the guts but providing additionally high-level wrapper for convenience.

**Comment by perrygeo** [comment]
Draft API via example:

```python
from mapbox import Uploader
u = Uploader('username')
url = u.stage('test.tif')  # uses boto3 and uploads API to push to s3
                           # option to pass in creds if you want to use your own s3 account

# power users with data already web-accessible, you start here
job = u.upload(url, 'test1')  # creates `username.test1`

# ... wait ....
print(u.status(job)['Complete'])  #True
```

Would love to hear thoughts on:
* passing around data versus maintaining state in the Uploader instance
* naming of classes and methods

I've got a basic implementation of the above API. Still need to write tests and such. PR coming soon...

**Comment by perrygeo** [comment]
@sgillies Thinking about what the methods should return, I see the geocoder returns raw responses. I was thinking of approaching it by returning the response dict but raising an exception if the response wasn't successful. 

    res = self.session.post(uri, json=msg)
    if res.status_code == 201:
        return res.json()
    else:
        raise Exception("Upload failed...")

I can see advantages to both approaches, seems to be a question of whether we want the sdk or the user to take responsibility for checking the status of the response.

**Comment by sgillies** [comment]
@perrygeo how about `Uploader.stage` and `Uploader.extract` for the fine grained methods and `Uploader.upload` for the coarse, do-it-all method?

**Comment by perrygeo** [comment]
> Uploader.upload would need to track credential-getting, staging, and extracting states, right?

Yes. Just pushed a new upload method that would do just that

However, we can't test the `stage` method (or the new `upload`) method until I can mock s3. My initial attempt with `moto` is pushed by commented out.

> how soon in development do we need to make this async?

Not sure. But luckily the `extract` step returns immediately and allows you to poll with `status`

**Comment by sgillies** [comment]
Closed via #31. More iteration on the uploads module in future tickets.
